### PR TITLE
Add Contract domain object, closes #229

### LINKED ISSUE #229: Create EDC `Contract` domain object
Hello everyone,

we are currently missing a `Contract` domain object. This object should represent a `ContractOffer` two parties have agreed upon. 

I'd like to use this issue to discuss how an initial implementation could look like and then create the corresponding PR for it.

What do you think about this?

```java
public interface Contract {
    @NotNull String getId();

    @NotNull URI getProvider();

    @NotNull URI getConsumer();

    @NotNull ZonedDateTime getContractSigningDate();

    @NotNull ZonedDateTime getContractStartDate();

    @NotNull ZonedDateTime getContractEndDate();

    @NotNull List<Asset> getAssets();

    @Nullable Policy getPolicy();

    @Nullable ContractDocument getContractDocument();
}

```

The `Policy` represents contract details in ODRL, whereas the `ContractDocument` may contain the text of a human readable contract document. I assume both properties are optional, but a correct contract should contain at least one of them.

---
Current State:

```java
public class Contract {
    @NotNull String getId();

    @NotNull URI getProvider();

    @NotNull URI getConsumer();

    @NotNull ZonedDateTime getContractSigningDate();

    @NotNull ZonedDateTime getContractStartDate();

    @NotNull ZonedDateTime getContractEndDate();

    @NotNull List<Asset> getAssets();

    @Nullable Policy getPolicy();
}

```

<sub>Dominik Pinsel <dominik.pinsel@daimler.com>, Daimler TSS GmbH, [legal info/Impressum](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md) </sub>

### PR DESCRIPTION
**Pull request by DominikPinsel**
Signed-off-by: Dominik Pinsel <dominik.pinsel@daimler.com>

<sub>Dominik Pinsel <dominik.pinsel@daimler.com>, Daimler TSS GmbH, [legal info/Impressum](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md) </sub>

### CODE REVIEW

**Comment by ndr-brt** [comment]
Better address this only on #244 ? Because this PR does just introduce a class that's not used by anyone

**Comment by DominikPinsel** [comment]
> Better address this only on #244 ? Because this PR does just introduce a class that's not used by anyone

There are probably multiple issues depend on the `Contract` class. So I wanted to create a very small, single purpose pull-request. Like this no issues, that depend on the `Contract`, are blocked.

File: spi/src/main/java/org/eclipse/dataspaceconnector/spi/types/domain/contract/Contract.java [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+ */
+public class Contract {
+
+    private final String id;
+    private final String providerId;
+    private final String consumerId;
+    private final long contractSigningDate;
+    private final long contractStartDate;
+    private final long contractEndDate;
+    private final List<Asset> assets;
```
Comment (DominikPinsel) [inline spi/src/main/java/org/eclipse/dataspaceconnector/spi/types/domain/contract/Contract.java:35]: change asset to asset id

File: spi/src/main/java/org/eclipse/dataspaceconnector/spi/types/domain/contract/Contract.java [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+import org.eclipse.dataspaceconnector.spi.types.domain.asset.Asset;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * {@link Contract} to regulate data transfer between two parties.
+ */
+public class Contract {
```
Comment (DominikPinsel) [inline spi/src/main/java/org/eclipse/dataspaceconnector/spi/types/domain/contract/Contract.java:27]: contract to contract agreement

File: spi/src/main/java/org/eclipse/dataspaceconnector/spi/types/domain/contract/Contract.java [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+
+    private final String id;
+    private final String providerId;
```
Comment (DominikPinsel) [inline spi/src/main/java/org/eclipse/dataspaceconnector/spi/types/domain/contract/Contract.java:30]: provider id to provider agent id

File: spi/src/main/java/org/eclipse/dataspaceconnector/spi/types/domain/contract/Contract.java [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    private final String consumerId;
```
Comment (DominikPinsel) [inline spi/src/main/java/org/eclipse/dataspaceconnector/spi/types/domain/contract/Contract.java:31]: consumer id to consumer agent id

**Review by jimmarino — APPROVED** [review]
LGTM, should I merge it as soon as the tests pass?

**Comment by DominikPinsel** [comment]
@jimmarino  I added all the changes from our review this afternoon. From my side this PR can be merged.

**Review by jimmarino — APPROVED** [review]

**Pull request merged**